intermediate_source/named_tensor_tutorial.py

```diff
@@ -0,0 +1,519 @@
+"""
+
+import torch
+imgs = torch.randn(1, 2, 2, 3 , names=('N', 'C', 'H', 'W'))
+print(imgs.names)
+
+######################################################################
+# Unlike in
+# `the original named tensors blogpost <http://nlp.seas.harvard.edu/NamedTensor>`_,
+# named dimensions are ordered: ``tensor.names[i]`` is the name of the ``i`` th dimension of ``tensor``.
+#
+# There are two ways rename a ``Tensor``'s dimensions:
+
+# Method #1: set .names attribute
+imgs.names = ['batch', 'channel', 'width', 'height']
+print(imgs.names)
+
+# Method #2: specify new names:
+imgs = imgs.rename(channel='C', width='W', height='H')
+print(imgs.names)
+
+######################################################################
+# The preferred way to remove names is to call ``tensor.rename(None)``:
+
+imgs = imgs.rename(None)
+print(imgs.names)
+
+######################################################################
+# Unnamed tensors (tensors with no named dimensions) still work as normal and do
+# not have names in their ``repr``.
+
+unnamed = torch.randn(2, 1, 3)
+print(unnamed)
+print(unnamed.names)
+
+######################################################################
+# Named tensors do not require that all dimensions be named.
+
+imgs = torch.randn(3, 1, 1, 2, names=('N', None, None, None))
+print(imgs.names)
+
+######################################################################
+# Because named tensors coexist with unnamed tensors, we need a nice way to
+# write named-tensor-aware code that works with both named and unnamed tensors.
+# Use ``tensor.refine_names(*names)`` to refine dimensions and lift unnamed dims
+# to named dims. Refining a dimension is defined as a "rename" with the following
+# constraints:
+#
+# - A ``None`` dim can be refined to have any name
+# - A named dim can only be refined to have the same name.
+
+imgs = torch.randn(3, 1, 1, 2)
+imgs = imgs.refine_names('N', 'C', 'H', 'W')
+print(imgs.names)
+
+# Coerces the last two dims to 'H' and 'W'. In Python 2, use the string '...' instead of ...
+print(imgs.refine_names(..., 'H', 'W').names)
+
+def catch_error(fn):
+    try:
+        fn()
+    except RuntimeError as err:
+        print(err)
+
+# Tried to refine an existing name to a different name
+catch_error(lambda: imgs.refine_names('batch', 'channel', 'height', 'width'))
+
+######################################################################
+# Most simple operations propagate names. The ultimate goal for named tensors is
+# for all operations to propagate names in a reasonable, intuitive manner. Many
+# common operations have been added at the time of the 1.3 release; here,
+# for example, is `.abs()`:
+
+named_imgs = imgs.refine_names('N', 'C', 'H', 'W')
+print(named_imgs.abs().names)
+
+######################################################################
+# Speaking of operations propogating names, let's quickly cover one
+# of the most important operations in PyTorch.

```

Comment:
Why is matrix multiply one of the "most important operations"? Let's not use value statements and instead leave it to the user to judge if an operator is important to them or not.

Also, can we move this back to the name inference section? It's important to see that there is a check names step and a propagate names step and to understand what matrix multiply does for both. For something like `abs`, that has really simple a name inference rule this isn't necessary.

Reply:
The original order of these sections was:
- Accessors / reduction
- Name inference
  - basic 1D example
  - broadcasting
  - matrix multiply
- explicit broadcasting by names

I understand that it is difficult to go automatic broadcasting to matrix multiply and then to explicit broadcasting by names, but we shouldn't introduce matrix multiply before name inference.

Reply:
Got it. I re-ordered this because I thought the broadcasting stuff all made sense together. I will order it back. Thanks!

Reply:
Reading through this again: I see what you were going for: broadcasting and matrix multiply are examples of where name inference can have unexpected semantics. 

I made broadcasting and matmul subheadings, rather than main headings, in the next version, to reinforce this.